File: src/main/java/cn/chahuyun/economy/manager/BackpackManager.java
package cn.chahuyun.economy.manager;

import cn.chahuyun.authorize.EventComponent;
import cn.chahuyun.authorize.MessageAuthorize;
import cn.chahuyun.authorize.constant.MessageMatchingEnum;
import cn.chahuyun.economy.entity.UserBackpack;
import cn.chahuyun.economy.entity.UserInfo;
import cn.chahuyun.economy.entity.props.UseEvent;
import cn.chahuyun.economy.exception.Operation;
import cn.chahuyun.economy.prop.PropBase;
import cn.chahuyun.economy.prop.PropsManager;
import cn.chahuyun.economy.utils.Log;
import cn.chahuyun.economy.utils.MessageUtil;
import lombok.val;
import net.mamoe.mirai.Bot;
import net.mamoe.mirai.contact.Group;
import net.mamoe.mirai.contact.Member;
import net.mamoe.mirai.event.events.GroupMessageEvent;
import net.mamoe.mirai.message.data.*;
import org.jetbrains.annotations.NotNull;

import java.util.Iterator;
import java.util.List;
import java.util.Objects;

/**
 * 背包管理
 *
 * @author Moyuyanli
 * @date 2022/11/15 10:00
 */
@EventComponent
public class BackpackManager {


    @MessageAuthorize(
            text = {"我的背包", "backpack"}
    )
    public void viewBackpack(GroupMessageEvent event) {
        Log.info("背包指令");

        Member sender = event.getSender();
        Bot bot = event.getBot();
        Group group = event.getSubject();

        UserInfo userInfo = UserManager.getUserInfo(sender);

        List<UserBackpack> backpacks = userInfo.getBackpacks();

        if (backpacks.isEmpty()) {
            group.sendMessage(MessageUtil.formatMessageChain(event.getMessage(), "你的背包为空!"));
            return;
        }

        var index = 1;
        var pageSize = 30;
        var totalSize = backpacks.size();
        var maxIndex = (int) Math.ceil((double) totalSize / pageSize);

        int fromIndex = 0;
        int toIndex = Math.min(index * pageSize, totalSize);
        var currentBackpacks = backpacks.subList(fromIndex, toIndex);
        showBackpack(bot, currentBackpacks, group, index, maxIndex);

        while (true) {
            GroupMessageEvent nextMessage = MessageUtil.INSTANCE.nextUserForGroupMessageEventSync(group.getId(), sender.getId(), 30);
            if (nextMessage == null || !nextMessage.getMessage().contentToString().matches("[上下]一页")) {
                return;
            }
            val string = nextMessage.getMessage().contentToString();
            boolean shouldUpdate = false;

            if (string.equals("上一页") && index > 1) {
                index--;
                shouldUpdate = true;
            } else if (string.equals("下一页") && index < maxIndex) {
                index++;
                shouldUpdate = true;
            }

            if (shouldUpdate) {
                fromIndex = (index - 1) * pageSize;
                toIndex = Math.min(index * pageSize, totalSize);
                currentBackpacks = backpacks.subList(fromIndex, toIndex);
                showBackpack(bot, currentBackpacks, group, index, maxIndex);
            }
        }
    }

    private static void showBackpack(Bot bot, List<UserBackpack> backpacks, Group group, int currentPage, int maxPage) {
        ForwardMessageBuilder iNodes = new ForwardMessageBuilder(group);
        iNodes.add(bot, new PlainText("以下是你的背包↓:"));

        for (UserBackpack backpack : backpacks) {
            PropBase prop = PropsManager.getProp(backpack);
            if (prop == null) {
                continue;
            }
            iNodes.add(bot, new PlainText(String.format("物品id:%d%n%s", backpack.getPropId(), prop)));
        }
        iNodes.add(bot, MessageUtil.formatMessage("--- 当前页数: %d / 最大页数: %d ---", currentPage, maxPage));
        group.sendMessage(iNodes.build());
    }

    @SuppressWarnings("ForLoopReplaceableByForEach")
    @MessageAuthorize(
            text = "use( \\d+)+|使用( \\d+)+",
            messageMatching = MessageMatchingEnum.REGULAR
    )
    public void useProp(GroupMessageEvent event) {
        Member sender = event.getSender();
        MessageChain message = event.getMessage();
        String content = message.contentToString();
        Group group = event.getSubject();

        UserInfo userInfo = UserManager.getUserInfo(sender);
        UseEvent useEvent = new UseEvent(sender, group, userInfo);


        String[] split = content.split(" ");

        MessageChainBuilder builder = new MessageChainBuilder();
        builder.add(new QuoteReply(message));
        builder.add("本次使用道具:");

        for (int i = 1; i < split.length; i++) {
            long propId = Long.parseLong(split[i]);


            List<UserBackpack> backpacks = userInfo.getBackpacks();

            boolean success = false;
            for (Iterator<UserBackpack> iterator = backpacks.iterator(); iterator.hasNext(); ) {
                UserBackpack backpack = iterator.next();
                if (backpack.getPropId().equals(propId)) {
                    PropBase prop = PropsManager.getProp(backpack);
                    String messageProp = "使用成功";
                    boolean remove = false;
                    try {
                        prop.use(useEvent);
                    } catch (Exception e) {
                        if (e instanceof Operation) {
                            Operation operation = (Operation) e;
                            remove = operation.isRemove();
                            messageProp = operation.getMessage();
                        } else {
                            throw e;
                        }
                    }
                    if (remove) {
                        delPropToBackpack(userInfo, backpack);
                    } else {
                        PropsManager.updateProp(backpack.getPropId(), prop);
                    }
                    builder.add(MessageUtil.formatMessage("\n%d %s!", propId, messageProp));
                    success = true;
                    break;
                }
            }
            if (!success) {
                builder.add(MessageUtil.formatMessage("\n%d 你没有这个道具!", propId));
            }
        }

        group.sendMessage(builder.build());
    }


    @MessageAuthorize(
            text = "dis( \\d+)+|丢弃( \\d+)+",
            messageMatching = MessageMatchingEnum.REGULAR
    )
    public void discard(GroupMessageEvent event) {
        Member sender = event.getSender();
        MessageChain message = event.getMessage();
        String content = message.contentToString();
        Group group = event.getSubject();

        String[] split = content.split(" ");

        MessageChainBuilder builder = new MessageChainBuilder();
        builder.add(new QuoteReply(message));
        builder.add("本次丢弃道具:");

        UserInfo userInfo = UserManager.getUserInfo(sender);

        for (int i = 1; i < split.length; i++) {
            long propId = Long.parseLong(split[i]);

            List<UserBackpack> backpacks = userInfo.getBackpacks();
            boolean match = false;
            for (UserBackpack backpack : backpacks) {
                if (backpack.getPropId().equals(propId)) {
                    PropBase prop = PropsManager.getProp(backpack);
                    String name = prop.getName();
                    delPropToBackpack(userInfo, propId);
                    builder.add(MessageUtil.formatMessage("\n你丢掉了你的 %s 。", name));
                    match = true;
                    break;
                }
            }

            if (!match) {
                builder.add(MessageUtil.formatMessage("\n没找到 %d 的道具。", propId));
            }
        }

        group.sendMessage(builder.build());
    }

    /**
     * 添加一个道具到背包
     *
     * @param userInfo 用户
     * @param id       道具id
     */
    public static void addPropToBackpack(UserInfo userInfo, String code, String kind, @NotNull Long id) {
        UserBackpack userBackpack = new UserBackpack(userInfo.getId(), code, kind, id);
        userInfo.addPropToBackpack(userBackpack);
    }

    /**
     * 给这个用户删除这个道具
     *
     * @param userInfo 用户
     * @param id       道具id
     */
    public static void delPropToBackpack(UserInfo userInfo, Long id) {
        List<UserBackpack> backpacks = userInfo.getBackpacks();

        UserBackpack find = null;
        for (UserBackpack backpack : backpacks) {
            if (backpack.getPropId().equals(id)) {
                find = backpack;
                break;
            }
        }

        userInfo.removePropInBackpack(find);

        PropsManager.destroyPros(id);
    }

    /**
     * 给这个用户删除这个道具
     *
     * @param userInfo     用户
     * @param userBackpack 道具
     */
    public static void delPropToBackpack(UserInfo userInfo, UserBackpack userBackpack) {
        userInfo.removePropInBackpack(userBackpack);
        PropsManager.destroyPros(userBackpack.getPropId());
    }

    /**
     * 检查该用户是否拥有这个道具
     *
     * @param userInfo 用户信息
     * @param id       道具id
     * @return true 拥有
     */
    public static boolean checkPropInUser(UserInfo userInfo, Long id) {
        for (UserBackpack backpack : userInfo.getBackpacks()) {
            if (Objects.equals(backpack.getPropId(), id)) {
                return true;
            }
        }
        return false;
    }

    /**
     * 检查该用户是否拥有这个道具
     *
     * @param userInfo 用户信息
     * @param code     道具code
     * @return true 拥有
     */
    public static boolean checkPropInUser(UserInfo userInfo, String code) {
        for (UserBackpack backpack : userInfo.getBackpacks()) {
            if (Objects.equals(backpack.getPropCode(), code)) {
                return true;
            }
        }
        return false;
    }

}

File: src/main/kotlin/cn/chahuyun/economy/manager/BackpackManager.kt
package cn.chahuyun.economy.manager

import cn.chahuyun.economy.data.proxy.EntityProxyRegistry
import cn.chahuyun.economy.model.user.UserBackpackDto
import cn.chahuyun.economy.model.user.UserInfoDto
import cn.chahuyun.economy.prop.BaseProp
import cn.chahuyun.economy.prop.PropsManager
import cn.chahuyun.economy.prop.Stackable
import cn.chahuyun.economy.utils.MessageUtil
import net.mamoe.mirai.Bot
import net.mamoe.mirai.contact.Group
import net.mamoe.mirai.message.data.ForwardMessageBuilder
import net.mamoe.mirai.message.data.PlainText

/**
 * 鑳屽寘绠＄悊鍣?
 * 鑳屽寘鐩稿叧鐨?闈炰簨浠剁洃鍚?閫昏緫銆?
 *
 * 璇存槑锛?
 * - `action.BackpackAction` 浠呬繚鐣欐寚浠ゅ叆鍙ｄ笌鍙傛暟瑙ｆ瀽銆?
 * - 閬撳叿澧炲垹鏌ャ€佽儗鍖呭唴瀹规覆鏌撶瓑鍙鐢ㄩ€昏緫涓嬫矇鍒拌繖閲屻€?
 */
object BackpackManager {

    /**
     * 鏄剧ず鐢ㄦ埛鑳屽寘鍐呭
     *
     * @param bot 鏈哄櫒浜哄疄渚?
     * @param backpacks 鐢ㄦ埛鑳屽寘閬撳叿鍒楄〃
     * @param group 缇ょ粍瀹炰緥
     * @param currentPage 褰撳墠椤电爜
     * @param maxPage 鏈€澶ч〉鐮?
     */
    suspend fun showBackpack(
        bot: Bot,
        backpacks: List<UserBackpackDto>,
        group: Group,
        currentPage: Int,
        maxPage: Int,
    ) {
        val nodes = ForwardMessageBuilder(group)
        nodes.add(bot, PlainText("浠ヤ笅鏄綘鐨勮儗鍖呪啌:"))

        // 閬嶅巻鑳屽寘涓殑閬撳叿骞舵坊鍔犲埌娑堟伅鑺傜偣涓?
        for (backpack in backpacks) {
            val prop = PropsManager.getProp(backpack) ?: continue
            nodes.add(bot, PlainText("鐗╁搧id:${backpack.propId}\n$prop"))
        }
        nodes.add(bot, MessageUtil.formatMessage("--- 褰撳墠椤垫暟: ${currentPage} / 鏈€澶ч〉鏁? ${maxPage} ---"))
        group.sendMessage(nodes.build())
    }

    /**
     * 娣诲姞涓€涓亾鍏峰埌鑳屽寘
     *
     * @param userInfo 鐢ㄦ埛淇℃伅
     * @param code 閬撳叿缂栫爜
     * @param kind 閬撳叿绫诲瀷
     * @param id 閬撳叿ID
     */
    @JvmStatic
    fun addPropToBackpack(userInfo: UserInfoDto, code: String, kind: String, id: Long): UserBackpackDto {
        require(id != 0L) { "涓嶈兘娣诲姞鏃犳晥閬撳叿鍒拌儗鍖? propId=$id, code=$code" }
        val userBackpack = UserBackpackDto(
            userId = userInfo.id,
            propCode = code,
            propKind = kind,
            propId = id
        )
        val saved = backpackProxy.save(userBackpack)
        if (saved.id == 0L || backpackProxy.findById(saved.id) == null) {
            error("淇濆瓨鑳屽寘璁板綍澶辫触: userId=${userInfo.id}, code=$code, propId=$id")
        }
        userInfo.backpacks = userInfo.backpacks + saved
        userInfo.backpackCount = userInfo.backpacks.size
        return saved
    }

    /**
     * 鍙戞斁鍙爢鍙犻亾鍏枫€傚凡鏈夊悓 code 閬撳叿鏃跺悎骞舵暟閲忥紝鍚﹀垯鍒涘缓鏂板疄渚嬪苟鍔犲叆鑳屽寘銆?     */
    @JvmStatic
    fun addStackablePropToBackpack(userInfo: UserInfoDto, code: String, kind: String, amount: Int): UserBackpackDto {
        require(amount > 0) { "鍙戞斁鏁伴噺蹇呴』澶т簬0: code=$code, amount=$amount" }

        userInfo.backpacks.find { it.propCode == code }?.let { backpack ->
            val prop = PropsManager.getProp(backpack)
                ?: error("鑳屽寘閬撳叿鏁版嵁涓嶅瓨鍦? code=$code, propId=${backpack.propId}")
            require(prop is Stackable && prop.isStack) { "鑳屽寘閬撳叿涓嶆槸鍙爢鍙犻亾鍏? code=$code" }
            prop.num += amount
            PropsManager.updateProp(backpack.propId, prop)
            return backpack
        }

        val prop = PropsManager.getTemplate(code, BaseProp::class.java)
        require(prop is Stackable && prop.isStack) { "閬撳叿妯℃澘涓嶆槸鍙爢鍙犻亾鍏? code=$code" }
        prop.num = amount
        val propId = PropsManager.addProp(prop)
        return addPropToBackpack(userInfo, code, kind, propId)
    }

    /**
     * 鏍规嵁閬撳叿ID鍒犻櫎鐢ㄦ埛鑳屽寘涓殑閬撳叿
     *
     * @param userInfo 鐢ㄦ埛淇℃伅
     * @param id 閬撳叿ID
     */
    @JvmStatic
    fun delPropToBackpack(userInfo: UserInfoDto, id: Long) {
        val backpacks = userInfo.backpacks
        val find = backpacks.find { it.propId == id }
        if (find != null) {
            backpackProxy.delete(find.id)
            userInfo.backpacks = userInfo.backpacks.filterNot { it.id == find.id }
            userInfo.backpackCount = userInfo.backpacks.size
            PropsManager.destroyPros(id)
        }
    }

    /**
     * 鏍规嵁UserBackpack瀵硅薄鍒犻櫎鐢ㄦ埛鑳屽寘涓殑閬撳叿
     *
     * @param userInfo 鐢ㄦ埛淇℃伅
     * @param userBackpack 鐢ㄦ埛鑳屽寘瀵硅薄
     */
    @JvmStatic
    fun delPropToBackpack(userInfo: UserInfoDto, userBackpack: UserBackpackDto) {
        backpackProxy.delete(userBackpack.id)
        userInfo.backpacks = userInfo.backpacks.filterNot { it.id == userBackpack.id }
        userInfo.backpackCount = userInfo.backpacks.size
        userBackpack.propId.takeIf { it != 0L }?.let { PropsManager.destroyPros(it) }
    }

    /**
     * 妫€鏌ョ敤鎴疯儗鍖呬腑鏄惁鍖呭惈鎸囧畾ID鐨勯亾鍏?
     *
     * @param userInfo 鐢ㄦ埛淇℃伅
     * @param id 閬撳叿ID
     * @return 濡傛灉鍖呭惈杩斿洖true锛屽惁鍒欒繑鍥瀎alse
     */
    @JvmStatic
    fun checkPropInUser(userInfo: UserInfoDto, id: Long): Boolean {
        return userInfo.backpacks.any { it.propId == id }
    }

    /**
     * 妫€鏌ョ敤鎴疯儗鍖呬腑鏄惁鍖呭惈鎸囧畾缂栫爜鐨勯亾鍏?
     *
     * @param userInfo 鐢ㄦ埛淇℃伅
     * @param code 閬撳叿缂栫爜
     * @return 濡傛灉鍖呭惈杩斿洖true锛屽惁鍒欒繑鍥瀎alse
     */
    @JvmStatic
    fun checkPropInUser(userInfo: UserInfoDto, code: String): Boolean {
        return userInfo.backpacks.any { it.propCode == code }
    }

    private val backpackProxy
        get() = EntityProxyRegistry.get<UserBackpackDto>("user_backpack") ?: error("背包代理器未初始化")
}


